core/transaction_validator.go

```diff
@@ -0,0 +1,51 @@
+// Copyright 2017 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package core
+
+import (
+	"errors"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/core/vm"
+)
+
+var (
+	ErrIncorrectPaygasMode = errors.New("incorrect PaygasMode for EVM")
+)
+
+func Validate(tx *types.Transaction, s types.Signer, evm *vm.EVM) error {
+	if evm.PaygasMode() != vm.PaygasHalt {
```

Comment:
Shouldn't this function take a gas limit? Since we'd want to terminate if it goes over say the 400k?

Reply:
Why do we need a signer here, btw?

Reply:
> Shouldn't this function take a gas limit? Since we'd want to terminate if it goes over say the 400k?

Ah of course, forgot about that. Will add.

Reply:
> Why do we need a signer here, btw?

That's because the existing `tx.AsMessage()` function expects a signer that it then passes on to `Sender()`. I could implement my own that just sets `from = ENTRY_POINT`, but that would then also skip e.g. the chain id checks (and would be unnecessary duplicate code).